### DISCUSSION TITLE
Functions to try parsing date formats

### DISCUSSION THREAD

**Discussion by hawkfish** [duckdb/duckdb#6959] [Ideas]
Date parsing is a [real bear](https://github.com/hawkfish/dateparse/blob/f1221546a3bf07d44b46a975ef0b5b6ba4bc0562/TR-2018-01.pdf). One thing that would help a lot is to be able to check whether a date parses with a given format without aborting the entire query (just like `try_cast` does). 

Even better would be to take a list of formats and try each one until you find one that works. This could be used by both the regular `strptime` and `try_strptime`.

It is important to let the user control the priority here as there can be ambiguities in date parsing. if we try to reorder the parsing, we could mess up the user's preferred disambiguation.

**Comment by FrancoisLepoutre** [comment]
Hi Richard, it looks like a pretty sound idea and, most important, a time saver on a pretty dull activity! But, sure, I have rarely found smartness in this field at db engine level. My experience is possibly a dated one though:-)

I have no sharp ideas on how to set this up in a "smart" way... But as far as I remember "hacking" this kind of stuff in a quick and dirty way at application level (not the best place) what you mentioned n your paper at Tableau Research makes sense:
1- the date testing order should be surely highly locale-dependant,
2- smart null shunting and decent sample size are important.

Our duck is a columnar engine, that should be helpful in making an educated guess!

**Reply by hawkfish** [reply]
Sadly I think I'd need a lawyer to determine whether I can try to recreate the functionality in DuckDB. 

But sampling single vectors is probably not a good idea because individual formats tend to cluster. Ideally you would want to use other data in the row to determine provenance and hence best guesses at formatting. To make that work in a relational model, you might want to use a [Divide](https://textform.readthedocs.io/en/latest/transforms/reshapes/divide.html) operation to route the strings into columns that are likely to be homogeneous based on the other fields. Then the `try_` variant can be used to parse and then use a [Merge](https://textform.readthedocs.io/en/latest/transforms/merges/merge.html) operation (probably `COALESE` in SQL) to stick them back together.
